File: fb/203.Remove-Linked-List-Elements.cpp
// Time O(n)
// Space O(1)
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode sentinel;
        sentinel.next = head;

        ListNode *prev = &sentinel, *curr = head, *toDelete = nullptr;
        while (curr != nullptr) {
          if (curr->val == val) {
            prev->next = curr->next;
            toDelete = curr;
          } else prev = curr;

          curr = curr->next;

          if (toDelete != nullptr) {
            delete toDelete;
            toDelete = nullptr;
          }
        }

        return sentinel.next;
    }
};

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

File: fb/203.Remove-Linked-List-Elements.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        fakeHead = ListNode(0)
        pointer = fakeHead
        
        while head:
            if head.val != val:
                pointer.next = ListNode(head.val)
                pointer = pointer.next
            head = head.next
            
        return fakeHead.next
